### DISCUSSION TITLE
How to distinguish None value sent by user and default None value for an optional argument in a API

### DISCUSSION THREAD

**Discussion by phuchm** [fastapi/fastapi#11899] [Questions]
### Privileged issue

- [X] I'm @tiangolo or he asked me directly to create an issue here.

### Issue Content

I have a API with 3 arguments, one of them is Optional with default value as None.
When user send a post request without this argument -> ok, continue to process this argument with None value.
But when user send None value for this argument, I want to detect and return error

**Comment by PhysicallyActive** [comment]
I'm not sure I understand what you're asking. Do you have three different pydantic-models as input arguments? Do you have one pydantic-model with three properties? Or do you have three query-parameters?

The following (using the "arguments" as query parameters) works as you expect:

```python
from fastapi import APIRouter, FastAPI

app = FastAPI()


@app.post("/hello")
async def hello_world(a: int, b: int, c: int | None = None) -> dict[str, str]:
    return {
        "message": f"Hello, World! The sum of {a} and {b} is {a + b}, {c} is optional"
    }
```

**Comment by PhysicallyActive** [comment]
If you instead wanted to prevent the user from inputting "none" into the body of a post-request, you can do it this way (note, that this is not a FastAPI issue, but a validation-constraint):

```python
import pydantic
from fastapi import APIRouter, FastAPI

app = FastAPI()


class Test(pydantic.BaseModel):
    a: int
    b: int
    c: int | None = None

    @pydantic.field_validator("c", mode="before")
    @classmethod
    def test_input(cls, v: int | None) -> int | None:
        if v is None:
            raise ValueError("You're not allowed to input None")
        return v


@app.post("/hello")
async def hello_world(test: Test) -> dict[str, str]:
    return {
        "message": f"Hello, World! The sum of {test.a} and {test.b} is {test.a + test.b}, {test.c} is optional"
    }
```

**Reply by phuchm** [reply]
I have three query parameters. What is the difference between one pydantic model with three properties and three query parameters? You mean that if using pydantic model, we can handle sending None value from user by pydantic.field_validator?

**Comment by binbjz** [comment]
You should move your question to the discussion area. This space is for tracking issues. There will be more people paying attention to your question there.
